query-parser/src/parse.rs

```diff
@@ -512,22 +517,42 @@ def_parser!(Find, aggregate_element, Element, {
 });
 
 def_parser!(Find, pull_element, Element, {
-    seq().of_exactly(Query::pull().with(Query::variable().and(Query::pull_attributes())))
+    Query::pull().with(Query::variable().and(Query::pull_attributes()))
          .map(|(var, attrs)| Element::Pull(Pull { var: var, patterns: attrs }))
 });
 
-def_parser!(Find, elem, Element, {
-    choice([
-        try(Find::variable_element()),
+enum ElementType {
+    Corresponding,
+    Pull,
+    Aggregate,
+}
+
+def_parser!(Find, seq_elem, Element, {
+    let element_parser_for_type = |ty: ElementType| {
+        match ty {
+            ElementType::Corresponding => Find::corresponding_element(),
+            ElementType::Pull => Find::pull_element(),
+            ElementType::Aggregate => Find::aggregate_element(),
+        }
+    };
 
+    // This slightly tortured phrasing ensures that we don't consume
```

Comment:
I'm glad I asked you to look at this, 'cuz I didn't find this pattern when I skimmed `combine` a while back.  I'll try it out in `tx-parser` and see if it's a good fit.  Thanks!